rme-what4/src/Data/RME/What4.hs

```diff
@@ -111,14 +152,81 @@ set s = M (\_ _ t -> t () s)
 -- | The state of evaluating an Expr into an RME term
 data S t = S
   { nextVar :: !Int -- ^ next fresh variable to be used with RME lit
-  , nonceCache :: !(MapF.MapF (Nonce.Nonce t) SomeR) -- ^ previously translated w4 expressions
+  , nonceCache :: !(MapF (Nonce.Nonce t) R') -- ^ previously translated w4 expressions
+  , abstracts :: !(MapF (AbstractKey t) R') -- ^ previously assigned function applications
   }
 
+-- | The application of a symbolic function to arguments in the RME domain.
+data AbstractKey t ret where
+  AbstractKey ::
+    Nonce.Nonce t (args ::> ret) ->
+    Assignment RMERepr args ->
+    RMERepr ret ->
+    Assignment R' args ->
+    AbstractKey t ret
+
+instance TestEquality (AbstractKey t) where
+  testEquality (AbstractKey f ts_ _ xs_) (AbstractKey g _ _ ys_) =
+    [Refl | Refl <- testEquality f g, Refl <- go ts_ xs_ ys_]
```

Comment:
It took me a while to figure out that this is defined using a monad comprehension where `Maybe` is the `Monad`. It might be worth leaving a brief comment saying as much. (Similarly for the `TestEquality (ConcreteKey t)` instance.)